lib/msal-browser/src/app/PublicClientApplication.ts

```diff
@@ -114,35 +122,50 @@ export class PublicClientApplication extends ClientApplication implements IPubli
         const silentRequestKey = JSON.stringify(thumbprint);
+
         const cachedResponse = this.activeSilentTokenRequests.get(silentRequestKey);
         if (typeof cachedResponse === "undefined") {
+            if (request.silentTokenRetrievalStrategy === SilentTokenRetrievalStrategy.CacheOnly) {
+                throw BrowserAuthError.createAccessTokenNotInCacheError();
+            }
+
             this.logger.verbose("acquireTokenSilent called for the first time, storing active request", correlationId);
-            const response = this.acquireTokenSilentAsync({
-                ...request,
-                correlationId
-            }, account)
-                .then((result) => {
-                    this.activeSilentTokenRequests.delete(silentRequestKey);
-                    atsMeasurement.endMeasurement({
-                        success: true,
-                        fromCache: result.fromCache,
-                        isNativeBroker: result.fromNativeBroker
-                    });
-                    atsMeasurement.flushMeasurement();
-                    return result;
-                })
-                .catch((error: AuthError) => {
-                    this.activeSilentTokenRequests.delete(silentRequestKey);
-                    atsMeasurement.endMeasurement({
-                        errorCode: error.errorCode,
-                        subErrorCode: error.subError,
-                        success: false
-                    });
-                    atsMeasurement.flushMeasurement();
-                    throw error;
-                });
-            this.activeSilentTokenRequests.set(silentRequestKey, response);
-            return response;
+
+            let tokenError;
+            let token: AuthenticationResult;
+            let measurement = {};
+            try {
+                token = await this.acquireTokenSilentAsync({
+                    ...request,
+                    correlationId,
+                }, account);
+                measurement = {
+                    success: true,
+                    fromCache: token.fromCache,
+                    isNativeBroker: token.fromNativeBroker,
+                };
+            } catch (error) {
+                tokenError = error as AuthError;
+                measurement = {
+                    success: false,
+                    errorCode: tokenError.errorCode,
+                    subErrorCode: tokenError.subError,
+                };
+            }
+
+            this.activeSilentTokenRequests.delete(silentRequestKey);
+            atsMeasurement.endMeasurement(measurement);
+            atsMeasurement.flushMeasurement();
+
+            if (tokenError) {
+                throw tokenError;
+            }
+
+            const tokenPromise = new Promise<AuthenticationResult>((resolve) => {
+                resolve(token);
+            });
+            this.activeSilentTokenRequests.set(silentRequestKey, tokenPromise);
+            return tokenPromise;
```

Comment:
Why is this refactored? Seems unrelated (and the usage of .then/.catch was intentional).

Reply:
And yeah, this isn't working the way the previous code was, as youll note by failing unit tests. Please revert, thanks!

Reply:
I could update the unit tests...

Reply:
It does not function the same. We use promise chain explicitly because we create the promise chain and then memorize it, so that if another request with the same request thumbprint is made, we return the previous promise for that request, instead of making a new request.

Reply:
See #3837. Async/await does not work here because `await` executes the promise, and returns the resolved value. We need a reference to the underlying promise, so that we can reuse it.

Reply:
We'll make sure there is a comment explaining this for future reference.

Reply:
I see, thanks for explaining.